src/hwperfmon.c

```diff
@@ -0,0 +1,356 @@
+/* Copyright 2020 SiFive, Inc */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#include <metal/hwperfmon.h>
+#include <stdint.h>
+
+/* Macro to generate code within a switch case */
+#define METAL_HWPERFMON_HANDLE_SWITCH(m)                                       \
+    m(3) m(4) m(5) m(6) m(7) m(8) m(9) m(10) m(11) m(12) m(13) m(14) m(15)     \
+        m(16) m(17) m(18) m(19) m(20) m(21) m(22) m(23) m(24) m(25) m(26)      \
+            m(27) m(28) m(29) m(30) m(31)
+
+/* Macro to set values into event selector register */
+#define METAL_HWPERFMON_SET_EVENT_REG(x)                                       \
+    case METAL_HWPERFMON_COUNTER_##x:                                          \
+        __asm__ __volatile__("csrr %0, mhpmevent" #x : "=r"(val));             \
+        val &= ~bitmask;                                                       \
+        val |= bitmask;                                                        \
+        __asm__ __volatile__("csrw mhpmevent" #x ", %0" : : "r"(val));         \
+        break;
+
+/* Macro to set values into event selector register */
+#define METAL_HWPERFMON_CLR_EVENT_REG(x)                                       \
+    case METAL_HWPERFMON_COUNTER_##x:                                          \
+        __asm__ __volatile__("csrr %0, mhpmevent" #x : "=r"(val));             \
+        val &= ~bitmask;                                                       \
+        __asm__ __volatile__("csrw mhpmevent" #x ", %0" : : "r"(val));         \
+        break;
+
+/* Macro to get values from event selector register */
+#define METAL_HWPERFMON_GET_EVENT_REG(x)                                       \
+    case METAL_HWPERFMON_COUNTER_##x:                                          \
+        __asm__ __volatile__("csrr %0, mhpmevent" #x : "=r"(val));             \
+        break;
+
+/* Macro to read HW performance monitor counter values */
+#if __riscv_xlen == 32
+#define METAL_HWPERFMON_GET_COUNT_REG(x)                                       \
+    case METAL_HWPERFMON_COUNTER_##x:                                          \
+        __asm__ __volatile__("csrr %0, mhpmcounter" #x "h" : "=r"(vh));        \
+        __asm__ __volatile__("csrr %0, mhpmcounter" #x : "=r"(vl));            \

```

Comment:
Is this magically atomic somehow? If not, then an overflow from the low 32 bits into the high 32 bits can cause
the wrong value to be returned. The usual fix for this is to use a loop which reads high, reads low, reads high until the two high reads return the same value:
```
uint32_t h1, l, h2;
do {
   h1 = read_hi();
   l = read_lo();
   h2 = read_hi();
} while (h1 != h2);
```

Reply:
Corrected.